paper/src/main/java/com/denizenscript/denizen/paper/events/PlayerPreparesGrindstoneCraftScriptEvent.java

```diff
@@ -0,0 +1,115 @@
+
+public class PlayerPreparesGrindstoneCraftScriptEvent extends BukkitScriptEvent implements Listener {
+
+    // <--[event]
+    // @Events
+    // player prepares grindstone craft <item>
+    //
+    // @Group Player
+    //
+    // @Location true
+    //
+    // @Plugin Paper
+    //
+    // @Triggers when a player prepares to grind an item.
+    //
+    // @Context
+    // <context.inventory> returns the InventoryTag of the grindstone inventory.
+    // <context.result> returns the ItemTag to be crafted.
+    //
+    // @Determine
+    // "RESULT:" + ItemTag to change the item that is crafted.
+    //
+    // @Player Always.
+    //
+    // @Warning The player doing the grinding is estimated and may be inaccurate.
+    //
+    // @Examples
+    // This example removes also the cursed binding_curse enchantment from the item.
+    // on player prepares grindstone craft item:
+    // - determine result:<context.result.with[remove_enchantments=binding_curse]>
+    //
+    // -->
+
+
+    public PlayerPreparesGrindstoneCraftScriptEvent() {
+        registerCouldMatcher("player prepares grindstone craft <item>");
+    }
+
+    public PrepareResultEvent event;
+    public ItemTag item;
+    public Inventory inventory;
+    public PlayerTag player;
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        if (!item.tryAdvancedMatcher(path.eventArgLowerAt(4))) {
+            return false;
+        }
+        if (!runInCheck(path, inventory.getLocation())) {
+            return false;
+        }
+        return super.matches(path);
+    }
+
+    @Override
+    public boolean applyDetermination(ScriptPath path, ObjectTag determinationObj) {
+        if (determinationObj instanceof ElementTag) {
+            String lower = CoreUtilities.toLowerCase(determinationObj.toString());
+            if (lower.startsWith("result:")) {
+                ItemTag result = ItemTag.valueOf(lower.substring("result:".length()), path.container);
+                event.setResult(result.getItemStack());
+                return true;
+            }
+        }
+        return super.applyDetermination(path, determinationObj);
+    }
+
+    @Override
+    public ObjectTag getContext(String name) {
+        switch (name) {
+            case "inventory": return InventoryTag.mirrorBukkitInventory(inventory);
+            case "result": return new ItemTag(event.getResult());
+        }
+        return super.getContext(name);
+    }
+
+    @Override
+    public ScriptEntryData getScriptEntryData() {
+        return new BukkitScriptEntryData(player, null);
+    }
+
+    @EventHandler
+    public void onPlayerPreparesGrindstoneCraft(PrepareResultEvent event) {
+        inventory = event.getInventory();
+        if (!(inventory instanceof GrindstoneInventory)) {
+            return;
+        }
+        if (inventory.getViewers().isEmpty()) {
+            return;
+        }
+        HumanEntity humanEntity = inventory.getViewers().get(0);

```

Comment:
If you do remove the `inventory` field, then can use `event.getViewers()` instead of `event.getInventory().getViewers()` - does the same thing but looks cleaner

Reply:
Yep. Done.